cfbs/analyze.py

```diff
@@ -0,0 +1,521 @@
+
+            for filepath in files_of_checksum:
+                if filepath in checksum_mpf_files_dict:
+                    # 1A1. a match of both checksum and filepath:
+                    for version in checksum_mpf_files_dict[filepath]:
+                        versions_data.version_counter.increment(version)
+
+                    filepaths_highest_versions[filepath] = highest_version(
+                        checksum_mpf_files_dict[filepath]
+                    )
+                else:
+                    # 1A2. there are files with the same checksum in MPF but not the same filepath:
+                    if filepath in mpf_files_dict:
+                        # 1A2A. filepath exists somewhere else but not for this checksum:
+                        filepath_versions = []
+                        for mpf_checksum in mpf_files_dict[filepath]:
+                            filepath_versions += mpf_files_dict[filepath][mpf_checksum]
+                        for version in filepath_versions:
+                            versions_data.different_filepath_vc.increment(version)
+                        dfv_fhv[filepath] = highest_version(filepath_versions)
+                    else:
+                        # 1A2B. checksum exists but filepath is not known:
+                        # there are no versions to count since the filepath is not known
+                        pass
+
+        for filepath in files_of_checksum:
+            if filepath in filepaths_highest_versions:
+                versions_data.highest_version_counter.increment(
+                    filepaths_highest_versions[filepath]
+                )
+            if filepath in dfv_fhv:
+                versions_data.different_filepath_hvc.increment(dfv_fhv[filepath])
+
+    if reference_version is None:
+        reference_version = versions_data.version_counter.most_common_version()
+
+    # if not a single file in the analyzed policyset has an MPF-known checksum, and a specific `reference_version` was not given, `reference_version` will still be `None`
+    if reference_version is None:
+        reference_version_files = []
+        reference_version_checksums = {}
+    else:
+        reference_version_files = mpf_versions_dict[reference_version].keys()
+        reference_version_checksums = {}
+        for mpf_filepath in mpf_versions_dict[reference_version]:
+            mpf_checksum = mpf_versions_dict[reference_version][mpf_filepath]
+            if mpf_checksum not in reference_version_checksums:
+                reference_version_checksums[mpf_checksum] = []
+            reference_version_checksums[mpf_checksum].append(mpf_filepath)
+
+    analyzed_files = AnalyzedFiles(reference_version)
+
+    # categorize all files, based on their relation with the reference version and known MPF files:
+    # 1. files present:
+    for checksum, files_of_checksum in checksums_dict.items():
+        if checksum in mpf_checksums_dict:
+            # 1A. checksum known:
```

Comment:
I am a little confused by the duplicate 1A, 1A1, 1A1A etc here. I know the comments aren't functional but the duplication seems like it would make it harder to refer/search to a particular section.

Reply:
Ideally, I think you could refactor each branch label into a function that is shared and that might improve the readability of this section which has quite a deep nest of indents which is often a sign of need for refactoring I would say.

Reply:
I think this is fine as-is. The code works and is 80% readable. A refactor of the same labels as functions would probably improve readability but at this point I don't think it is critical at all.